File: examples/native-tasks-android/app/src/main/kotlin/com/pyreon/FlowEdgesPreview.kt
package com.pyreon

import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.pyreon.runtime.PyreonFlowEdgeCanvas
import com.pyreon.runtime.PyreonFlowEdgeSegment
import com.pyreon.runtime.PyreonFlowEdgeStroke
import com.pyreon.runtime.PyreonFlowViewport
import com.pyreon.runtime.PyreonFlowEdge
import com.pyreon.runtime.PyreonFlowNode
import com.pyreon.runtime.PyreonFlowState
import com.pyreon.runtime.PyreonFlowView
import com.pyreon.runtime.PyreonXYPosition

// Compiles PyreonFlowEdgeCanvas into the real app module — the first time any
// Gradle build has compiled the Kotlin canvas composable (it was declared
// kotlinSdkOnly and referenced by no device app). Not routed to; the build is
// the point. See FlowEdgesPreview.swift for the iOS twin.
@Composable
fun FlowEdgesPreview() {
    val strokes = listOf(
        PyreonFlowEdgeStroke(
            id = "e1",
            segments = listOf(
                PyreonFlowEdgeSegment.move(150.0, 20.0),
                PyreonFlowEdgeSegment.cubic(200.0, 20.0, c1x = 175.0, c1y = 20.0, c2x = 175.0, c2y = 20.0),
            ),
            color = "#0a8f7c",
            width = 2.0,
        ),
    )
    PyreonFlowEdgeCanvas(edges = strokes, viewport = PyreonFlowViewport(), modifier = Modifier.height(80.dp))
}

// Compiles the complete interactive host as part of the real Android app,
// including state-driven edge derivation and node content.
@Composable
fun FlowHostPreview() {
    val state = remember { PyreonFlowState(
        nodes = listOf(
            PyreonFlowNode("start", position = PyreonXYPosition(0.0, 0.0), data = "Start"),
            PyreonFlowNode("end", position = PyreonXYPosition(200.0, 80.0), data = "End"),
        ),
        edges = listOf(PyreonFlowEdge("edge", source = "start", target = "end")),
    ) }
    PyreonFlowView(
        state = state,
        modifier = Modifier.width(400.dp).height(240.dp),
    ) { node ->
        Text(node.data)
    }
}

File: examples/native-tasks-ios/ios/FlowEdgesPreview.swift
import SwiftUI

// Compiles PyreonFlowEdgeCanvas into the real app target. Nothing in shared
// source can reference it yet (no <Flow> host emit), so without this file the
// canvas is linked by no device build at all — "the device gate covers it"
// was true of nothing. A hand-wired SwiftUI view is exactly how the runtime
// documents consuming the canvas today; this one is small enough to be the
// reference for that. Not routed to from the app; the point is the build.
@available(iOS 17.0, *)
struct FlowEdgesPreview: View {
    let strokes: [PyreonFlowEdgeStroke] = [
        PyreonFlowEdgeStroke(
            id: "e1",
            segments: [.move(150, 20), .cubic(200, 20, c1x: 175, c1y: 20, c2x: 175, c2y: 20)],
            color: "#0a8f7c",
            width: 2),
    ]
    var body: some View {
        // `.equatable()` joins once #3320 (Equatable canvas) is on main.
        PyreonFlowEdgeCanvas(edges: strokes, viewport: PyreonFlowViewport(x: 0, y: 0, zoom: 1))
            .frame(height: 80)
    }
}
